### PR TITLE
[SPARK-35726][SQL] Truncate java.time.Duration by fields of day-time interval type

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Support truncate java.time.Duration by fields of day-time interval type.

### Why are the changes needed?
To respect fields of the target day-time interval types.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added UT

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala [hunk 1]
```diff
@@ -273,6 +274,24 @@ class CatalystTypeConvertersSuite extends SparkFunSuite with SQLHelper {
...
+      DayTimeIntervalType(DAY, MINUTE) -> 90060000000L,
+      DayTimeIntervalType(DAY, SECOND) -> 90061000000L,
+      DayTimeIntervalType(HOUR, HOUR) -> 90000000000L,
+      DayTimeIntervalType(HOUR, MINUTE) -> 90060000000L,
+      DayTimeIntervalType(HOUR, SECOND) -> 90061000000L,
+      DayTimeIntervalType(MINUTE, MINUTE) -> 90060000000L,
+      DayTimeIntervalType(MINUTE, SECOND) -> 90061000000L,
+      DayTimeIntervalType(SECOND, SECOND) -> 90061000000L)
+      .foreach { case (dt, value) =>
+        assert(CatalystTypeConverters.createToCatalystConverter(dt)(duration) == value)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala:290]: Could you check negative duration, please. Just add this:
```scala
        assert(CatalystTypeConverters.createToCatalystConverter(dt)(duration.negated()) == -value)
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala [hunk 2]
```diff
@@ -892,17 +892,35 @@ object IntervalUtils {
...
-    } else {
-      val microsInSeconds = Math.multiplyExact(seconds, MICROS_PER_SECOND)
-      Math.addExact(microsInSeconds, duration.getNano / NANOS_PER_MICROS)
+    endField match {
+      case DayTimeIntervalType.DAY =>
+        secondsToMicros(seconds - seconds % SECONDS_PER_DAY)
+      case DayTimeIntervalType.HOUR =>
+        secondsToMicros(seconds - seconds % SECONDS_PER_HOUR)
+      case DayTimeIntervalType.MINUTE =>
+        secondsToMicros(seconds - seconds % SECONDS_PER_MINUTE)
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala:921]: hmm, what about nanos part?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala [hunk 3]
```diff
@@ -273,6 +274,24 @@ class CatalystTypeConvertersSuite extends SparkFunSuite with SQLHelper {
     assert(errMsg.contains("long overflow"))
   }
 
+  test("SPARK-35726: Truncate java.time.Duration by fields of day-time interval type") {
+    val duration = Duration.ofSeconds(90061)
```
Comment (MaxGekk) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala:278]: Could you test a Duration with nanos part.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala]: updated

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/IntervalUtils.scala]: updtaed

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala (hunk 3 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala]: done

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala (hunk 1 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/CatalystTypeConvertersSuite.scala]: the different is right according to the origin logic?
```
 (DayTimeIntervalType(SECOND, SECOND), 90062000000L, -90062000001L))
```

**Review by MaxGekk — APPROVED** [review]
+1, LGTM. Merging to master.
Thank you, @AngersZhuuuu .

**Pull request closed without merging**